src/OrleansAzureUtils/Providers/Storage/AzureBlobStorage.cs

```diff
@@ -261,5 +223,36 @@ private static string GetBlobName(string grainType, GrainReference grainId)
                 throw;
             }
         }
+
+        private async Task WriteStateAndCreateContainerIfNotExists(string grainType, GrainReference grainId, IGrainState grainState, string json, CloudBlockBlob blob)
+        {
+            try
+            {
+                await DoConditionalUpdate(() => blob.UploadTextAsync(json, Encoding.UTF8, AccessCondition.GenerateIfMatchCondition(grainState.ETag), null, null),
+                    blob.Name, grainState.ETag).ConfigureAwait(false);
+            }
+            catch (StorageException exception) when (exception.IsContainerNotFound())
+            {
+                // if the container does not exist, create it, and make another attempt
+                if (this.Log.IsVerbose3) this.Log.Verbose3((int)AzureProviderErrorCode.AzureBlobProvider_ContainerNotFound, "Creating container: GrainType={0} Grainid={1} ETag={2} to BlobName={3} in Container={4}", grainType, grainId, grainState.ETag, blob.Name, container.Name);
+                await container.CreateIfNotExistsAsync().ConfigureAwait(false);
+
+                await WriteStateAndCreateContainerIfNotExists(grainType, grainId, grainState, json, blob).ConfigureAwait(false);
```

Comment:
return WriteStateAndCreateContainerIfNotExists...

Reason is that after exception we duplicate work.  While this is ok now, because it's just a duplication of the same state change that was just done in the successful WriteStateAndCreateContainerIfNotExists, it may not always be ok.  For instance, what would happen if someone started tracking a successful write statistic at the end of this call.